File: python/check_stack_driver.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
import argparse
import requests
import json
import os
from google.cloud import monitoring_v3
from google.cloud.monitoring_v3.query import Query
import datetime

# 引数の解析
parser = argparse.ArgumentParser()
parser.add_argument("--project")
parser.add_argument("--instance")
parser.add_argument("--critical")
parser.add_argument("--warning")
parser.add_argument("--key")
args = parser.parse_args()

# 環境変数
os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = args.key

# GCP SDK
client = monitoring_v3.MetricServiceClient()
name = client.project_path(args.project)

# 値を取得期間を指定する
# 取得しようとしている値は最長 240 秒値が格納されないことがあるので、5分以前の値を取得する
# https://cloud.google.com/monitoring/api/metrics_gcp#gcp-compute
endtime = datetime.datetime.utcnow()
endtime = endtime - datetime.timedelta(minutes=5)

# メトリック取得クエリの発行
# https://googleapis.dev/python/monitoring/latest/query.html#google.cloud.monitoring_v3.query.Query
query = Query(
            client,
            args.project,
            'compute.googleapis.com/instance/uptime', # 取得するメトリック
            end_time=endtime,
            days=0,
            hours=0,
            minutes=int(args.critical), # 取得する分数を指定する
        )

# 指定したインスタンスのみに値を絞る
query = query.select_metrics(instance_name=args.instance) 

uptime = 0.0
for timeseries in query:
    for point in timeseries.points:
        uptime += point.value.double_value # UPTIMEを合算し、起動時間とする

# 起動時間が0の時は起動していない、CRITICALエラーをだす
if (uptime <= 0.0) :
    print("CRITICAL - Instance {} stopped for {} minutes.".format(args.instance, args.critical))
    exit(2)

# 起動時間がWARNINGしきい値未満の場合、WARNINGエラーをだす
if (uptime <= int(args.warning) * 60) :
    print("WARNING - Instance {} stopped for {} minutes.".format(args.instance, args.warning))
    exit(1)

print("OK - Instance {} is UP.".format(args.instance))
exit(0)

File: go/check_stack_driver.go
package main

import (
	"os"
	"fmt"
	"strconv"
	flags "github.com/jessevdk/go-flags"
	"time"
	"google.golang.org/api/iterator"
	"golang.org/x/net/context"
	monitoring "cloud.google.com/go/monitoring/apiv3"
	googlepb "github.com/golang/protobuf/ptypes/timestamp"
	monitoringpb "google.golang.org/genproto/googleapis/monitoring/v3"
)

type Options struct {
	Project  string `short:"p" long:"project"  description:"gcp project id"       required:"true"`
	Instance string `short:"i" long:"instance" description:"gce instance name"    required:"true"`
	Critical string `short:"c" long:"critical" description:"critical threshold"   required:"true"`
  	Warning  string `short:"w" long:"warning"  description:"warning threshold"    required:"true"`
  	Auth     string `short:"a" long:"auth"     description:"gcp authenticate key" required:"true"`
}

func main() {
    // 引数解析処理
    var opts Options
  	parser := flags.NewParser(&opts, flags.IgnoreUnknown)
  	_, err := parser.Parse()
    if err != nil {
		fmt.Println("Missing required arguments.")
        parser.WriteHelp(os.Stdout)
        os.Exit(1)
  	}

	// 定数定義
	os.Setenv("GOOGLE_APPLICATION_CREDENTIALS", opts.Auth)

	// GCPのSDKを使う準備
	// https://godoc.org/cloud.google.com/go/monitoring/apiv3#NewMetricClient
	ctx := context.Background()
	c, err := monitoring.NewMetricClient(ctx)
	if err != nil {
		// TODO: Handle error.
	}
	
	// フィルタの生成
	var filter string = "metric.type = \"compute.googleapis.com/instance/uptime\" "
	filter += "AND metric.label.instance_name = \"" + opts.Instance + "\" "

	// 取得期間の生成
	// https://godoc.org/google.golang.org/genproto/googleapis/monitoring/v3#ListTimeSeriesRequest
	var critical int64
	critical, _ = strconv.ParseInt(opts.Critical, 10, 64)
	var warning int64
	warning, _ = strconv.ParseInt(opts.Warning, 10, 64)
	unixNow := time.Now().Unix()
	req := &monitoringpb.ListTimeSeriesRequest{
		Name : "projects/" + opts.Project,
		Filter : filter,
		Interval: &monitoringpb.TimeInterval{
			EndTime: &googlepb.Timestamp{
				Seconds: unixNow - 300,
			},
			StartTime: &googlepb.Timestamp{
				Seconds: unixNow - (critical * 60),
			},
		},
	}

	uptime := 0.0
	it := c.ListTimeSeries(ctx, req)
	for {
		resp, err := it.Next()
		if err == iterator.Done {
			break
		}
		if err != nil {
			// TODO: Handle error.
		}
		for _, point := range resp.Points {
			uptime += point.GetValue().GetDoubleValue()
		}
		_ = resp
	}

	if (int64(uptime) <=  0) {
		output(2, "CRITICAL - Instance " + opts.Instance + " stopped for " + opts.Critical + " minutes.")
	}

	if (int64(uptime) <=  warning * 60) {
		output(1, "WARNING - Instance " + opts.Instance + " stopped for " + opts.Critical + " minutes.")
	}

	output(0, "OK - Instance " + opts.Instance + " is UP.")
}

func output(status int, message string) {
	fmt.Println(message)
	os.Exit(status)
}